Instance actions

jaredm-ibm opened the pull request:
Fixes general issues with the way body payloads were generated.
Adds stop, start, reboot actions on the instance.
Also adds status test methods and a wait_for method to instance.
wait_for blocks until instance is in either stopped or started states depending on parameters.

agrare commented on lib/ibm/cloud/sdk/vpc/instances.rb:
```diff
@@ -38,21 +38,111 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+          def wait_for(started: true, sleep_time: 5, timeout: 600)
+            refresh
+            # When in transitional state wait for it to be stable.
+            until test_state(started)
```
We could generalize this a bit if we accepted a block from the caller so: `wait_for(&:stopped?)` which makes more intuitive sense than `wait_for(started:false)`.

This way we can support any target state (e.g. not sure if VPC supports suspending or a suspended state) without adding more boolean arguments.  Might be a moot point if the only options are running or stopped.

jaredm-ibm replied:
Hah, yeah, I was thinking of adding a block over the weekend. I remembered that the PureApp version took a block which was useful. It would be nice to have the option to modify the logic without updating the method. I can look at it later. Using the boolean was to avoid typos and reduce complexity when setting the state to look for.

agrare reviewed (approved):
Looks good, up to you if you want to update wait_for to take a block here or in a follow-up

jaredm-ibm commented:
I'm going to reset this to WIP while I get the plugin actions working. I want to move some logging here maybe reimplement wait_for with a block.

jaredm-ibm commented:
* Add logging to wait_for and action create methods.
* Added ability to use block in wait_for.
* Added method to switch from apache to curl http output. Apache output is very terse and curl output is verbose. There is a potentiality of leaking details when using curl so apache is default.

jaredm-ibm commented:
@agrare This should be ready for merge. I have the actions finished in the plugin in and doubt there will be any more changes.
The block is nice.

```
      instance = vpc.instances.instance(ems_ref)
      instance.actions.start
      instance.wait_for do
        update!(:raw_power_state => instance.status)
        instance.started?
      end
    end

```

agrare commented on lib/ibm/cloud/sdk/vpc/instances.rb:
```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
 
         # Work with a single instance.
         class Instance < VPCInstance
+          TRANSITIONAL_STATES = %(pausing pending restarting resuming starting stopping)
```
I think you intended for this to be an array of strings:

```
>> %(pausing pending restarting resuming starting stopping)
=> "pausing pending restarting resuming starting stopping"
>> %w[pausing pending restarting resuming starting stopping]
=> ["pausing", "pending", "restarting", "resuming", "starting", "stopping"]
```

What you have here has the problem of:
```
>> %(pausing pending restarting resuming starting stopping).include?("pausing pending")
=> true
```

agrare commented on lib/ibm/cloud/sdk/vpc/instances.rb:
```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+          TRANSITIONAL_STATES = %(pausing pending restarting resuming starting stopping)
+          ERROR_STATE = 'failed'
+          RUNNING_STATE = 'running'
+          STOPPED_STATES = %('stopped', 'paused')
```
Same here, should be `%w[stopped paused]`

jaredm-ibm replied:
Good catch, thanks

agrare commented on lib/ibm/cloud/sdk/vpc/instances.rb:
```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+
+          # Used with wait_for return true if started is true and vm is started, or started is false and vm is stopped.
+          # @return [Boolean]
+          def test_state(started: true, &block)
```
Hm combining test_state and a block seems to be extra complexity when you can accomplish everything with
```
def wait_for(sleep_time: 5, timeout: 600, &block)
  loop do
    refresh
    raise "VM #{id} is in a failed state." if failed?
    break if block.call(self)
    timeout = sleep_counter(sleep_time, timeout)
    raise "Time out while waiting #{id} to be stable." if timeout <= 0
  end
end
```

And the difference to the caller in the simple case is `wait_for(started: true)` vs `wait_for(&:started?)`

I won't hold this up for this but I think we could simplify this a lot without losing anything

jaredm-ibm replied:
I was debating on removing started and had pretty much the same syntax. I don't like having options that can be silently overridden by other options. I deferred to the more complicated since using started allows people less familiar with blocks to use wait_for for the simplest reasons can just pass in an option. I guess another option is to add two more methods wait_for_started and wait_for_stopped each passing in the requisite block in its body. That is probably a better solution than the started option. I'll update.

agrare commented on lib/ibm/cloud/sdk/vpc/instances.rb:
```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+            refresh
+            @logger.info("Testing state of instance #{id} with status #{status}")
+            if block_given?
+              return true if block.call
```
Can you yield `self` so that the caller can do e.g. `wait_for(&:started?)`

agrare commented on lib/ibm/cloud/sdk/vpc/instances.rb:
```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+              raise "Time out while waiting #{id} to be stable." if timeout <= 0
+            end
+            @logger.info("Finished wait for instance #{id} ends in state #{status}.")
+            test_state(&block)
```
Is the point of this last block to return if the wait was successful or not?  Seems like it will either be successful or it'll raise a timeout.
Also if we don't pass `started` and we weren't called with a block wouldn't a call to `wait_for(started: false)` end up calling `test_state` with `started: true`?

jaredm-ibm replied:
Uggg, yeah, not including started is a bug.

jaredm-ibm replied:
The point of the last is to have a boolean return. Originally I was returning false when an error was hit, but that didn't seem like enough information to determine why so I raised. It probably could have no return since anything that isn't true will raise, and thus the return is useless. I'll remove the last call.

agrare replied:
Yeah I like the return on success raise on any error its a pretty common pattern.  Typically it is denoted by `!` in the method name so `wait_for!(&:something)` raises an exception and `wait_for(&:something)` returns true/false.
Don't think we need both version but we could name this `wait_for!` so it is obvious that it'll raise on a failure

jaredm-ibm replied:
Hah, we could have both versions if we returned a tuple with (boolean, reason) in the standard version and then acted accordingly in the ! version. But that seems like overkill for the moment.

agrare reviewed (approved):
Looks awesome!

agrare commented:
Released in v0.1.11
